### PR TITLE
Ephemeral access tokens

### PR DESCRIPTION
**Pull request by ab77**
@klutchell this PR should be ready to go, please review when you get a chance. There are some structural changes in here, especially around tests.

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -79,25 +79,57 @@ jobs:
...
+      INSTALLATION_ID: "${{ vars.INSTALLATION_ID || 34040165 }}"
+      TOKEN_SCOPE: >-
+        {
+          "administration": "write"
+        }
       pr_branch: "flowzone/${{ github.head_ref || github.ref_name }}"
       pr_title: "Test Flowzone PR ${{ github.event.pull_request.number }}"
 
     steps:
+      - name: Check secrets
```
Comment (ab77) [inline .github/workflows/tests.yml:95]: unfortunately we can't output the token from previous job, since it's masked, so have to duplicate code here.

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -151,52 +210,41 @@ jobs:
...
+          filter_re: "^policy-?bot"
+          GH_TOKEN: ${{ steps.gh_token.outputs.token || secrets.FLOWZONE_TOKEN }}
         run: |
-          while true
-          do
-            sleep 30
-
-            prev_checks="${checks}"
-
-            checks="$(gh pr status --json statusCheckRollup,number --jq '[.createdBy[] |
```
Comment (ab77) [inline .github/workflows/tests.yml:172]: `.createdBy` is empty when queried using ephemeral tokens. Must be a GH permissions thing requiring a token tied to a user (not machine/app) identity. No way around it as far as I can tell. I've re-implemented using `gh pr checks`, which is a bit kludgy but seems to work).

**Comment by klutchell** [comment]
@ab77 I hope to review this soon, I just have to wrap up some other tasks!

File: README.md [hunk 3]
```diff
@@ -330,11 +334,17 @@ but they can also be [configured for personal repositories](https://docs.github.
 
 These secrets can also be found at the top of [flowzone.yml](./.github/workflows/flowzone.yml).
 
+#### `PRIVATE_KEY`
```
Comment (Page-) [inline README.md:337]: I'd prefer a more specific name than `PRIVATE_KEY`, eg `GH_APP_PRIVATE_KEY`
```suggestion
#### `GH_APP_PRIVATE_KEY`
```
or maybe adding the ephemeral bit in there as well could make sense?

File: flowzone.yml [hunk 4]
```diff
@@ -109,11 +119,15 @@
     run: gh release delete --yes '${{ github.event.pull_request.head.ref }}' || true
     env:
       <<: *gitHubCliEnvironment
+      GH_TOKEN: ${{ steps.gh_token.outputs.token || secrets.FLOWZONE_TOKEN }}
```
Comment (Page-) [inline flowzone.yml:122]: This is repeated a lot, it could make sense to use anchors/aliases, eg
```suggestion
      <<:
        - *gitHubCliEnvironment
        - *ghTokenEnv
```
```yaml
- &ghTokenEnv
  GH_TOKEN: ${{ steps.gh_token.outputs.token || secrets.FLOWZONE_TOKEN }}

Comment (klutchell) [reply flowzone.yml:122]: +1

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -65,39 +65,96 @@ jobs:
...
+          fi
+
+          if [[ -n '${{ secrets.PRIVATE_KEY }}' ]]; then
+              echo 'available=true' >> $GITHUB_OUTPUT
+          else
+              echo 'available=false' >> $GITHUB_OUTPUT
+          fi
+
+      - name: Generate token
+        uses: tibdex/github-app-token@v1
```
Comment (klutchell) [inline .github/workflows/tests.yml:136]: We should pin this to a full semver since it's not an official github action.

Once this merges Renovate will pin it to a commit hash, and keep the version as a comment for tracking.

File: flowzone.yml [hunk 6]
```diff
@@ -41,7 +54,6 @@
     with:
       fetch-depth: 0
       submodules: "recursive"
-      token: ${{ secrets.FLOWZONE_TOKEN }}
```
Comment (klutchell) [inline flowzone.yml:44]: Why were these removed? If the intent is to use `${{ secrets.GITHUB_TOKEN }}` instead I would rather be clear about that rather than appear to have forgotten to specify.

File: flowzone.yml [hunk 7]
```diff
@@ -75,10 +86,10 @@
     with:
       registry: ghcr.io
       username: ${{ env.GHCR_USER || github.actor }}
-      password: ${{ secrets.GHCR_TOKEN || secrets.FLOWZONE_TOKEN }}
+      password: ${{ steps.gh_token.outputs.token || secrets.FLOWZONE_TOKEN }}
```
Comment (klutchell) [inline flowzone.yml:89]: If we are deprecating this secret we need to be clear about that. I thought it was preferable to allow the option for a different token with package write access. Maybe that's not required, but we should be clear that the secret is unused going forward.

File: flowzone.yml [hunk 8]
```diff
@@ -92,7 +103,6 @@
   - &gitHubCliEnvironment
     # environment variables used by gh CLI
     # https://cli.github.com/manual/gh_help_environment
-    GH_TOKEN: "${{ secrets.FLOWZONE_TOKEN }}"
```
Comment (klutchell) [inline flowzone.yml:95]: I think it would make more sense to provide the token here
```
GH_TOKEN: ${{ steps.gh_token.outputs.token || secrets.FLOWZONE_TOKEN }}
```
and just ensure we only call this anchor from the step context, and not the job. I think the only place we set it at the job context is protect branch anyway.

That avoids having to have two aliases whenever we need to set gh env.

File: flowzone.yml (hunk 4 above)
Comment (klutchell) [reply flowzone.yml:122]: or as I suggested above, keep this line in `*gitHubCliEnvironment`

File: flowzone.yml [hunk 9]
```diff
@@ -188,7 +206,45 @@ on:
...
         required: false
+
     inputs:
+      app_id:
+        description: "GitHub App id to impersonate"
+        type: string
+        required: false
+        # https://github.com/organizations/product-os/settings/apps/flowzone-app
+        # https://github.com/organizations/product-os/settings/variables/actions/APP_ID
+        default: "${{ vars.APP_ID || 291899 }}"
```
Comment (klutchell) [inline flowzone.yml:217]:
```suggestion
        default: "${{ vars.APP_ID || '291899' }}"
```

File: flowzone.yml [hunk 10]
```diff
@@ -188,7 +206,45 @@ on:
...
+      # not needed if installed on this current org/repo
+      installation_id:
+        description: "GitHub App installation id"
+        type: string
+        required: false
+        # https://github.com/organizations/product-os/settings/installations
+        # https://github.com/organizations/product-os/settings/variables/actions/INSTALLATION_ID
+        default: "${{ vars.INSTALLATION_ID || 34040165 }}"
```
Comment (klutchell) [inline flowzone.yml:225]:
```suggestion
        default: "${{ vars.INSTALLATION_ID || '34040165' }}"
```

File: flowzone.yml [hunk 11]
```diff
@@ -188,7 +206,45 @@ on:
...
+            "members": "read",
+            "metadata": "read",
+            "organization_secrets": "read",
+            "packages": "write",
+            "pages": "write",
+            "pull_requests": "read",
+            "secrets": "read",
+            "statuses": "read",
+            "workflows": "read"
+          }
```
Comment (klutchell) [inline flowzone.yml:247]: Could we require this to be base64 encoded so it's less likely to get mangled in a workflow file?

File: flowzone.yml [hunk 12]
```diff
@@ -457,6 +513,22 @@ jobs:
     steps:
       - *logGitHubContext
 
+      # https://stackoverflow.com/a/70249520/1559300
+      - name: Check secrets
```
Comment (klutchell) [inline flowzone.yml:517]: I think this got moved to the wrong job during a rebase. I would expect this step to be part of sanitize inputs.

File: .github/workflows/tests.yml (hunk 2 above)
Comment (klutchell) [reply .github/workflows/tests.yml:172]: I like how much cleaner it is, but I think `gh pr checks --required` will never pass if we change the name of a job.

These draft PRs intentionally don't update branch protections, so renamed jobs will be pending forever.

Comment (ab77) [reply .github/workflows/tests.yml:172]: @klutchell you'll notice that we always filter the output and don't use the exit code from GH CLI.

File: flowzone.yml (hunk 4 above)
Comment (ab77) [reply flowzone.yml:122]: @klutchell there is only place where `gitHubCliEnvironment` is used at the job/env level, so `steps` won't work there.

File: flowzone.yml (hunk 6 above)
Comment (ab77) [reply flowzone.yml:44]: That was the intention.

File: flowzone.yml (hunk 7 above)
Comment (ab77) [reply flowzone.yml:89]: I don't think there is a goods reason to keep it, I'll clean up the docs.

File: flowzone.yml (hunk 8 above)
Comment (ab77) [reply flowzone.yml:95]: See, https://github.com/product-os/flowzone/blob/master/flowzone.yml#L2531-L2533

File: .github/workflows/tests.yml [hunk 13]
```diff
@@ -151,52 +210,41 @@ jobs:
...
-            jq -r '.[].status' <<< "${checks}" | grep -q '^PENDING$' && continue
-
-            if [ "${prev_checks}" == "${checks}" ]
-            then
-              [ $(jq 'length' <<< "${checks}") -gt 0 ] || exit 1
-              jq -r '.[].status' <<< "${checks}" | grep -q '^COMPLETED$' && exit 0
-            fi
+          while ! gh pr checks --required \
+            --repo ${{ matrix.repository }} \
+            ${{ steps.cpr.outputs.pull-request-number}} | grep -qE "${flowzone_re}"; do
```
Comment (klutchell) [inline .github/workflows/tests.yml:223]: Is it okay if one Flowzone check has loaded, but not the others? I think visually I have seen some required checks take longer to show up than others, and I wonder if we wont end up in another race condition with GitHub.

Maybe I'm just paranoid from being burned too many times.

File: flowzone.yml (hunk 11 above)
Comment (ab77) [reply flowzone.yml:247]: @klutchell I would prefer this to be left as-is so it's clear what the token scope is. Looking at base64 strings isn't the best UX.

File: flowzone.yml (hunk 6 above)
Comment (klutchell) [reply flowzone.yml:44]: In my other PR I removed all permissions from this token at the workflow level. So this will break on private repos since we removed read access.

Fortunately you can just set new permissions at the versioned_source and tagged_source jobs to get what you need. Or change the default at the top of the workflow and keep docker_test with empty permissions.

File: flowzone.yml (hunk 8 above)
Comment (klutchell) [reply flowzone.yml:95]: Yeah, just move that to the steps instead of the job, so you can use a single alias.

File: flowzone.yml (hunk 11 above)
Comment (klutchell) [reply flowzone.yml:247]: fair enough

File: .github/workflows/tests.yml (hunk 13 above)
Comment (ab77) [reply .github/workflows/tests.yml:223]: @klutchell let's catch up on this - it's a wider issue I think than just this PR. You are right, checks registration lags.

File: .github/workflows/tests.yml [hunk 14]
```diff
@@ -151,52 +210,41 @@ jobs:
...
+              sleep $(((RANDOM % 5) + 5))
+          done
+          sleep $(((RANDOM % 15) + 15))
 
-            continue
+          while gh pr checks --required \
+            --repo ${{ matrix.repository }} \
+            ${{ steps.cpr.outputs.pull-request-number}} \
+            | grep -vE "${filter_re}" \
+            | grep -E '\s+pending\s+'; do
```
Comment (klutchell) [inline .github/workflows/tests.yml:233]: I think ~~if~~ when we change the name of a required job we will get stuck here.

Because you are filtering on required, and looping if they are still pending. These PRs don't update the required checks to match the new names, so a renamed/refactored job will be missing and pending forever.

Comment (klutchell) [reply .github/workflows/tests.yml:233]: Maybe if that's the case we can force merge the Flowzone PR, but it will come up at some point.

File: README.md [hunk 15]
```diff
@@ -25,11 +25,11 @@ Reusable, opinionated, zero-conf workflows for GitHub actions
   - [Website](#Website)
 - [Customization](#customization)
   - [Secrets](#secrets)
+    - [`GH_APP_PRIVATE_KEY`](#gh_app_private_key)
     - [`FLOWZONE_TOKEN`](#flowzone_token)
     - [`GPG_PRIVATE_KEY`](#gpg_private_key)
     - [`GPG_PASSPHRASE`](#gpg_passphrase)
     - [`NPM_TOKEN`](#npm_token)
-    - [`GHCR_TOKEN`](#ghcr_token)
```
Comment (klutchell) [inline README.md:32]: I assume no one is using this in production anyway 👍

Comment (ab77) [reply README.md:32]: Yeah, and if someone has, it's a small change to refactor their workflows.

File: .github/workflows/tests.yml (hunk 14 above)
Comment (ab77) [reply .github/workflows/tests.yml:233]: .. and we also have these regexes in `parse_prepare_protection_rules`. Probably can leave it as is for now and if it ever comes up, we can look at parametrising the names?

Comment (klutchell) [reply .github/workflows/tests.yml:233]: SGTM, maybe open a GH issue to investigate later

**Comment by ab77** [comment]
lgtm

**Comment by klutchell** [comment]
🎉
